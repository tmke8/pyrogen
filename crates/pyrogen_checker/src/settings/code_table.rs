use std::{
    f32::consts::E,
    fmt::{Debug, Display},
};

use pyrogen_macros::CacheKey;
use serde::{Deserialize, Serialize};

use crate::registry::{ErrorCode, ErrorCodeSet, ErrorCodeSetIterator};

/// A table to keep track of which error codes are enabled.
#[derive(Debug, CacheKey, Default)]
pub struct ErrorCodeTable {
    /// Maps rule codes to a boolean indicating if the rule should be autofixed.
    enabled: ErrorCodeSet,
    warning: ErrorCodeSet,
}

#[derive(Debug, PartialEq, Eq, Clone, Copy, Serialize, Deserialize)]
pub enum MessageKind {
    Error,
    Warning,
}

impl Display for MessageKind {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            MessageKind::Error => write!(f, "error"),
            MessageKind::Warning => write!(f, "warning"),
        }
    }
}

impl ErrorCodeTable {
    /// Creates a new empty error code table.
    pub const fn empty() -> Self {
        Self {
            enabled: ErrorCodeSet::empty(),
            warning: ErrorCodeSet::empty(),
        }
    }

    /// Returns whether the given rule should be checked.
    #[inline]
    pub const fn enabled(&self, rule: ErrorCode) -> bool {
        self.enabled.contains(rule)
    }

    #[inline]
    pub const fn entry(&self, rule: ErrorCode) -> Option<MessageKind> {
        if self.enabled(rule) {
            if self.is_warning(rule) {
                Some(MessageKind::Warning)
            } else {
                Some(MessageKind::Error)
            }
        } else {
            None
        }
    }

    /// Returns whether any of the given rules should be checked.
    #[inline]
    pub const fn any_enabled(&self, rules: &[ErrorCode]) -> bool {
        self.enabled
            .intersects(&ErrorCodeSet::from_error_codes(rules))
    }

    /// Returns whether violations of the given rule should be a warning.
    #[inline]
    pub const fn is_warning(&self, rule: ErrorCode) -> bool {
        self.warning.contains(rule)
    }

    /// Returns an iterator over all enabled rules.
    pub fn iter_enabled(&self) -> ErrorCodeSetIterator {
        self.enabled.iter()
    }

    /// Returns an iterator over all warnings.
    pub fn iter_warnings(&self) -> ErrorCodeSetIterator {
        self.warning.iter()
    }

    /// Enables the given rule.
    #[inline]
    pub fn enable_error(&mut self, rule: ErrorCode) {
        self.enabled.insert(rule);
    }

    /// Enables the given rule.
    #[inline]
    pub fn enable_warning(&mut self, rule: ErrorCode) {
        self.enabled.insert(rule);
        self.warning.insert(rule);
    }

    /// Disables the given rule.
    #[inline]
    pub fn disable(&mut self, rule: ErrorCode) {
        self.enabled.remove(rule);
        self.warning.remove(rule);
    }
}

impl FromIterator<ErrorCode> for ErrorCodeTable {
    fn from_iter<T: IntoIterator<Item = ErrorCode>>(iter: T) -> Self {
        let rules = ErrorCodeSet::from_iter(iter);
        Self {
            enabled: rules,
            warning: ErrorCodeSet::empty(),
        }
    }
}
