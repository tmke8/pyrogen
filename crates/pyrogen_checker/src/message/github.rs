use std::io::Write;

use crate::fs::relativize_path;
use crate::message::{Emitter, Message};
use crate::settings::code_table::MessageKind;

/// Generate error workflow command in GitHub Actions format.
/// See: [GitHub documentation](https://docs.github.com/en/actions/reference/workflow-commands-for-github-actions#setting-an-error-message)
#[derive(Default)]
pub struct GithubEmitter;

impl Emitter for GithubEmitter {
    fn emit(&mut self, writer: &mut dyn Write, messages: &[Message]) -> anyhow::Result<()> {
        for message in messages {
            let source_location = message.compute_start_location();
            let location = source_location.clone();

            let end_location = message.compute_end_location();
            let kind: &str = match message.kind {
                MessageKind::Error => "error",
                MessageKind::Warning => "warning",
            };

            write!(
                writer,
                "::{kind} title=Pyrogen \
                         ({code}),file={file},line={row},col={column},endLine={end_row},endColumn={end_column}::",
                kind = kind,
                code = message.diagnostic.error_code,
                file = message.filename(),
                row = source_location.row,
                column = source_location.column,
                end_row = end_location.row,
                end_column = end_location.column,
            )?;

            writeln!(
                writer,
                "{path}:{row}:{column}: {code} {body}",
                path = relativize_path(message.filename()),
                row = location.row,
                column = location.column,
                code = message.diagnostic.error_code,
                body = message.diagnostic.body,
            )?;
        }

        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use insta::assert_snapshot;

    use crate::message::tests::{capture_emitter_output, create_messages};
    use crate::message::GithubEmitter;

    #[test]
    fn output() {
        let mut emitter = GithubEmitter;
        let content = capture_emitter_output(&mut emitter, &create_messages());

        assert_snapshot!(content);
    }
}
