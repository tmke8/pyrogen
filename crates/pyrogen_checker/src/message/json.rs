use std::io::Write;

use serde::ser::SerializeSeq;
use serde::{Serialize, Serializer};
use serde_json::{json, Value};

use pyrogen_source_file::SourceCode;
use rustpython_ast::Ranged;

use crate::message::{Emitter, Message};

#[derive(Default)]
pub struct JsonEmitter;

impl Emitter for JsonEmitter {
    fn emit(&mut self, writer: &mut dyn Write, messages: &[Message]) -> anyhow::Result<()> {
        serde_json::to_writer_pretty(writer, &ExpandedMessages { messages })?;

        Ok(())
    }
}

struct ExpandedMessages<'a> {
    messages: &'a [Message],
}

impl Serialize for ExpandedMessages<'_> {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        let mut s = serializer.serialize_seq(Some(self.messages.len()))?;

        for message in self.messages {
            let value = message_to_json_value(message);
            s.serialize_element(&value)?;
        }

        s.end()
    }
}

pub(crate) fn message_to_json_value(message: &Message) -> Value {
    let source_code = message.file.to_source_code();

    let start_location = source_code.source_location(message.start());
    let end_location = source_code.source_location(message.end());
    let type_ignore_location = source_code.source_location(message.ignore_offset);

    json!({
        "code": message.diagnostic.error_code.to_string(),
        "message": message.diagnostic.body,
        "location": start_location,
        "end_location": end_location,
        "filename": message.filename(),
        "type_ignore_row": type_ignore_location.row,
        "kind": message.kind.to_string()
    })
}

#[cfg(test)]
mod tests {
    use insta::assert_snapshot;

    use crate::message::tests::{capture_emitter_output, create_messages};
    use crate::message::JsonEmitter;

    #[test]
    fn output() {
        let mut emitter = JsonEmitter;
        let content = capture_emitter_output(&mut emitter, &create_messages());

        assert_snapshot!(content);
    }
}
