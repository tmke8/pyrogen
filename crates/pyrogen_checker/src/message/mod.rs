use std::cmp::Ordering;
use std::collections::BTreeMap;
use std::io::Write;
use std::ops::Deref;

use rustc_hash::FxHashMap;
use rustpython_parser::ast::Ranged;
use rustpython_parser::text_size::{TextRange, TextSize};

use pyrogen_source_file::{SourceFile, SourceLocation};

pub use github::GithubEmitter;
pub use json::JsonEmitter;
pub use text::TextEmitter;

use crate::registry::{Diagnostic, DiagnosticKind};
use crate::settings::code_table::MessageKind;

// mod diff;
mod github;
mod json;
mod text;

#[derive(Debug, PartialEq, Eq)]
pub struct Message {
    pub diagnostic: DiagnosticKind,
    pub range: TextRange,
    pub file: SourceFile,
    pub ignore_offset: TextSize,
    pub kind: MessageKind,
}

impl Message {
    pub fn from_diagnostic(
        diagnostic: Diagnostic,
        file: SourceFile,
        ignore_offset: TextSize,
        kind: MessageKind,
    ) -> Self {
        Self {
            range: diagnostic.range(),
            diagnostic: diagnostic.kind,
            file,
            ignore_offset,
            kind,
        }
    }

    pub fn filename(&self) -> &str {
        self.file.name()
    }

    pub fn compute_start_location(&self) -> SourceLocation {
        self.file.to_source_code().source_location(self.start())
    }

    pub fn compute_end_location(&self) -> SourceLocation {
        self.file.to_source_code().source_location(self.end())
    }
}

impl Ord for Message {
    fn cmp(&self, other: &Self) -> Ordering {
        (&self.file, self.start()).cmp(&(&other.file, other.start()))
    }
}

impl PartialOrd for Message {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

impl Ranged for Message {
    fn range(&self) -> TextRange {
        self.range
    }
}

struct MessageWithLocation<'a> {
    message: &'a Message,
    start_location: SourceLocation,
}

impl Deref for MessageWithLocation<'_> {
    type Target = Message;

    fn deref(&self) -> &Self::Target {
        self.message
    }
}

fn group_messages_by_filename(messages: &[Message]) -> BTreeMap<&str, Vec<MessageWithLocation>> {
    let mut grouped_messages = BTreeMap::default();
    for message in messages {
        grouped_messages
            .entry(message.filename())
            .or_insert_with(Vec::new)
            .push(MessageWithLocation {
                message,
                start_location: message.compute_start_location(),
            });
    }
    grouped_messages
}

/// Display format for a [`Message`]s.
///
/// The emitter serializes a slice of [`Message`]'s and writes them to a [`Write`].
pub trait Emitter {
    /// Serializes the `messages` and writes the output to `writer`.
    fn emit(&mut self, writer: &mut dyn Write, messages: &[Message]) -> anyhow::Result<()>;
}

#[cfg(test)]
mod tests {
    use rustc_hash::FxHashMap;
    use rustpython_parser::ast::Ranged;
    use rustpython_parser::text_size::{TextRange, TextSize};

    use pyrogen_source_file::SourceFileBuilder;

    use crate::message::{Emitter, Message};
    use crate::registry::{Diagnostic, DiagnosticKind, ErrorCode};
    use crate::settings::code_table::MessageKind;

    pub(super) fn create_messages() -> Vec<Message> {
        let fib = r#"import os


def fibonacci(n):
    """Compute the nth number in the Fibonacci sequence."""
    x = 1
    if n == 0:
        return 0
    elif n == 1:
        return 1
    else:
        return fibonacci(n - 1) + fibonacci(n - 2)
"#;

        let unused_import = Diagnostic::new(
            DiagnosticKind {
                error_code: ErrorCode::UnusedImport,
                body: "`os` imported but unused".to_string(),
            },
            TextRange::new(TextSize::from(7), TextSize::from(9)),
        );

        let fib_source = SourceFileBuilder::new("fib.py", fib).finish();

        let unused_variable = Diagnostic::new(
            DiagnosticKind {
                error_code: ErrorCode::UnusedVariable,
                body: "Local variable `x` is assigned to but never used".to_string(),
            },
            TextRange::new(TextSize::from(94), TextSize::from(95)),
        );

        let file_2 = r#"if a == 1: pass"#;

        let undefined_name = Diagnostic::new(
            DiagnosticKind {
                error_code: ErrorCode::UndefinedName,
                body: "Undefined name `a`".to_string(),
            },
            TextRange::new(TextSize::from(3), TextSize::from(4)),
        );

        let file_2_source = SourceFileBuilder::new("undef.py", file_2).finish();

        let unused_import_start = unused_import.start();
        let unused_variable_start = unused_variable.start();
        let undefined_name_start = undefined_name.start();
        vec![
            Message::from_diagnostic(
                unused_import,
                fib_source.clone(),
                unused_import_start,
                MessageKind::Warning,
            ),
            Message::from_diagnostic(
                unused_variable,
                fib_source,
                unused_variable_start,
                MessageKind::Warning,
            ),
            Message::from_diagnostic(
                undefined_name,
                file_2_source,
                undefined_name_start,
                MessageKind::Error,
            ),
        ]
    }

    pub(super) fn capture_emitter_output(
        emitter: &mut dyn Emitter,
        messages: &[Message],
    ) -> String {
        let mut output: Vec<u8> = Vec::new();
        emitter.emit(&mut output, messages).unwrap();

        String::from_utf8(output).expect("Output to be valid UTF-8")
    }
}
