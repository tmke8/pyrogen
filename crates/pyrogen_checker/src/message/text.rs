use std::borrow::Cow;
use std::fmt::{Display, Formatter};
use std::io::Write;

use annotate_snippets::display_list::{DisplayList, FormatOptions};
use annotate_snippets::snippet::{Annotation, AnnotationType, Slice, Snippet, SourceAnnotation};
use bitflags::bitflags;
use colored::Colorize;

use pyrogen_source_file::OneIndexed;
use rustpython_parser::text_size::{TextRange, TextSize};

use crate::fs::relativize_path;
use crate::line_width::{LineWidthBuilder, TabSize};
use crate::message::{Emitter, Message};
use crate::registry::AsErrorCode;
use crate::settings::code_table::MessageKind;

bitflags! {
    #[derive(Default)]
    struct EmitterFlags: u8 {
        /// Whether to show the fix status of a diagnostic.
        const SHOW_FIX_STATUS = 0b0000_0001;
        /// Whether to show the diff of a fix, for diagnostics that have a fix.
        const SHOW_FIX_DIFF   = 0b0000_0010;
        /// Whether to show the source code of a diagnostic.
        const SHOW_SOURCE     = 0b0000_0100;
    }
}

#[derive(Default)]
pub struct TextEmitter {
    flags: EmitterFlags,
}

impl TextEmitter {
    #[must_use]
    pub fn with_show_fix_status(mut self, show_fix_status: bool) -> Self {
        self.flags
            .set(EmitterFlags::SHOW_FIX_STATUS, show_fix_status);
        self
    }

    #[must_use]
    pub fn with_show_fix_diff(mut self, show_fix_diff: bool) -> Self {
        self.flags.set(EmitterFlags::SHOW_FIX_DIFF, show_fix_diff);
        self
    }

    #[must_use]
    pub fn with_show_source(mut self, show_source: bool) -> Self {
        self.flags.set(EmitterFlags::SHOW_SOURCE, show_source);
        self
    }
}

impl Emitter for TextEmitter {
    fn emit(&mut self, writer: &mut dyn Write, messages: &[Message]) -> anyhow::Result<()> {
        for message in messages {
            write!(
                writer,
                "{path}{sep}",
                path = relativize_path(message.filename()).bold(),
                sep = ":".cyan(),
            )?;

            let start_location = message.compute_start_location();

            let diagnostic_location = start_location;

            writeln!(
                writer,
                "{row}{sep}{col}{sep} {code_and_body}",
                row = diagnostic_location.row,
                col = diagnostic_location.column,
                sep = ":".cyan(),
                code_and_body = RuleCodeAndBody { message }
            )?;

            if self.flags.intersects(EmitterFlags::SHOW_SOURCE) {
                writeln!(writer, "{}", MessageCodeFrame { message })?;
            }
        }

        Ok(())
    }
}

pub(super) struct RuleCodeAndBody<'a> {
    pub(crate) message: &'a Message,
}

impl Display for RuleCodeAndBody<'_> {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        let diagnostic = &self.message.diagnostic;

        match &self.message.kind {
            MessageKind::Error => {
                write!(
                    f,
                    "error: {body} [{code}]",
                    code = diagnostic.error_code().to_string().red().bold(),
                    body = diagnostic.body,
                )
            }
            MessageKind::Warning => {
                write!(
                    f,
                    "warn: {body} [{code}]",
                    code = diagnostic.error_code().to_string().yellow().bold(),
                    body = diagnostic.body,
                )
            }
        }
    }
}

pub(super) struct MessageCodeFrame<'a> {
    pub(crate) message: &'a Message,
}

impl Display for MessageCodeFrame<'_> {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        let Message {
            diagnostic: kind,
            file,
            range,
            ..
        } = self.message;

        let footer = Vec::new();

        let source_code = file.to_source_code();

        let content_start_index = source_code.line_index(range.start());
        let mut start_index = content_start_index.saturating_sub(2);

        // Trim leading empty lines.
        while start_index < content_start_index {
            if !source_code.line_text(start_index).trim().is_empty() {
                break;
            }
            start_index = start_index.saturating_add(1);
        }

        let content_end_index = source_code.line_index(range.end());
        let mut end_index = content_end_index
            .saturating_add(2)
            .min(OneIndexed::from_zero_indexed(source_code.line_count()));

        // Trim trailing empty lines.
        while end_index > content_end_index {
            if !source_code.line_text(end_index).trim().is_empty() {
                break;
            }

            end_index = end_index.saturating_sub(1);
        }

        let start_offset = source_code.line_start(start_index);
        let end_offset = source_code.line_end(end_index);

        let source = replace_whitespace(
            source_code.slice(TextRange::new(start_offset, end_offset)),
            range - start_offset,
        );

        let start_char = source.text[TextRange::up_to(source.annotation_range.start())]
            .chars()
            .count();

        let char_length = source.text[source.annotation_range].chars().count();

        let label = kind.error_code().to_string();

        let snippet = Snippet {
            title: None,
            slices: vec![Slice {
                source: &source.text,
                line_start: start_index.get(),
                annotations: vec![SourceAnnotation {
                    label: &label,
                    annotation_type: AnnotationType::Error,
                    range: (start_char, start_char + char_length),
                }],
                // The origin (file name, line number, and column number) is already encoded
                // in the `label`.
                origin: None,
                fold: false,
            }],
            footer,
            opt: FormatOptions {
                #[cfg(test)]
                color: false,
                #[cfg(not(test))]
                color: colored::control::SHOULD_COLORIZE.should_colorize(),
                ..FormatOptions::default()
            },
        };

        writeln!(f, "{message}", message = DisplayList::from(snippet))
    }
}

fn replace_whitespace(source: &str, annotation_range: TextRange) -> SourceCode {
    let mut result = String::new();
    let mut last_end = 0;
    let mut range = annotation_range;
    let mut line_width = LineWidthBuilder::new(TabSize::default());

    for (index, c) in source.char_indices() {
        let old_width = line_width.get();
        line_width = line_width.add_char(c);

        if matches!(c, '\t') {
            // SAFETY: The difference is a value in the range [1..TAB_SIZE] which is guaranteed to be less than `u32`.
            #[allow(clippy::cast_possible_truncation)]
            let tab_width = (line_width.get() - old_width) as u32;

            if index < usize::from(annotation_range.start()) {
                range += TextSize::new(tab_width - 1);
            } else if index < usize::from(annotation_range.end()) {
                range = range.add_end(TextSize::new(tab_width - 1));
            }

            result.push_str(&source[last_end..index]);

            for _ in 0..tab_width {
                result.push(' ');
            }

            last_end = index + 1;
        }
    }

    // No tabs
    if result.is_empty() {
        SourceCode {
            annotation_range,
            text: Cow::Borrowed(source),
        }
    } else {
        result.push_str(&source[last_end..]);
        SourceCode {
            annotation_range: range,
            text: Cow::Owned(result),
        }
    }
}

struct SourceCode<'a> {
    text: Cow<'a, str>,
    annotation_range: TextRange,
}

#[cfg(test)]
mod tests {
    use insta::assert_snapshot;

    use crate::message::tests::{capture_emitter_output, create_messages};
    use crate::message::TextEmitter;

    #[test]
    fn default() {
        let mut emitter = TextEmitter::default().with_show_source(true);
        let content = capture_emitter_output(&mut emitter, &create_messages());

        assert_snapshot!(content);
    }

    #[test]
    fn fix_status() {
        let mut emitter = TextEmitter::default()
            .with_show_fix_status(true)
            .with_show_source(true);
        let content = capture_emitter_output(&mut emitter, &create_messages());

        assert_snapshot!(content);
    }
}
