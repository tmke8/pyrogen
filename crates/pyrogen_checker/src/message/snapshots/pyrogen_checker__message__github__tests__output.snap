---
source: crates/pyrogen_checker/src/message/github.rs
expression: content
---
::warning title=Pyrogen (unused-import),file=fib.py,line=1,col=8,endLine=1,endColumn=10::fib.py:1:8: unused-import `os` imported but unused
::warning title=Pyrogen (unused-variable),file=fib.py,line=6,col=5,endLine=6,endColumn=6::fib.py:6:5: unused-variable Local variable `x` is assigned to but never used
::error title=Pyrogen (undefined-name),file=undef.py,line=1,col=4,endLine=1,endColumn=5::undef.py:1:4: undefined-name Undefined name `a`

